design/ceph-osd-redesign-for-octopus.md

```diff
@@ -0,0 +1,342 @@
+
+
+### Proposed Rook operator design
+The existing OSD operator design uses a two-stage process. The first stage is preparation of OSDs on
+a per-host basis, and the second stage is running individual OSDs. The design proposed here
+separates OSD tasks into more stages.
+
+A Rook OSD prepare binary currently handles the equivalent of stages 1 through 4, and in early
+implementations of the new OSD design, a version of the current binary could be used for stages 1
+through 3 to achieve the minimum viable product more quickly. Splitting the stages can be done
+afterwards as makes sense.
+
+#### Stage 1: Host device inventory
+The first stage is an inspection stage that is currently done as part of the existing preparation
+stage. Similar to the current Rook OSD prepare binary, this proposal calls for a Rook inspection
+binary which will examine the disk hardware, rejecting disks that are already in use by applications
+other than Rook (as per current behavior). Rook currently uses `lsblk` and custom logic to find free
+disks, but this design proposes to use `ceph-volume inventory` to find free disks. The inspection
+binary stores the result of the `ceph-volume inventory` command to a ConfigMap which  will be read
+by the operator.
+
+#### Stage 2: Process device inspection report
+In the second stage, the operator examines the host device inspection report in the ConfigMap. The
+primary inspection processing done would be to filter the list of devices reported based on the
+`deviceFilter` and the host `devices` specified in the cluster config.
+
+The insertion of stage 2 executed within the operator itself, placed in between stages 1 and 3
+keeps the majority of the Ceph configuration logic in the operator itself rather than distributing
+the logic between multiple binaries. Existing logic distribution has contributed to operation which
+is not always transparent or easy to extend. More on this below.
+
+#### Stage 3: Host device preparation
+This stage is one in which `ceph-volume` provisions the devices. Unless a custom Rook binary is
+absolutely necessary, `ceph-volume` can be called directly from the pod's container entrypoint.
+Arguments to `ceph-volume` will be based on the host devices found in stage 1 and processing done
+in stage 2.
+
+As a specific example, adding new OSD preparation options like support for device encryption
+involves changes to the Ceph Cluster CRD (required), to the OSD prepare pod specification created by
+the Ceph operator, to the Ceph command entrypoint for OSDs, and to the OSD preparation daemon
+itself. Splitting the preparation binary into 4 smaller, operator-centric steps means that changes
+to the Ceph Cluster CRD only require changes to the CRD and to the pods created by the operator,
+cutting the number of locations where code is modified in half.
+
+#### Stage 4: Host device OSD inspection
+Rook can determine which OSDs are prepared on the host by running `ceph-volume lvm list` in the same
```

Comment:
Why don't we refresh that map with the new content of the `ceph-volume inventory` command?

Reply:
Does `inventory` report the `lvm` information?

Reply:
Actually no, it does not, sorry!